text/0000-simd-infrastructure.md

```diff
@@ -0,0 +1,422 @@
+
+There is a new attribute: `repr(simd)`.
+
+```rust
+#[repr(simd)]
+struct f32x4(f32, f32, f32, f32);
+
+#[repr(simd)]
+struct Simd2<T>(T, T);
+```
+
+The `simd` `repr` can be attached to a struct and will cause such a
+struct to be compiled to a SIMD vector. It can be generic, but it is
+required that any fully monomorphised instance of the type consist of
+only a single "primitive" type, repeated some number of times. Types
+are flattened, so, for `struct Bar(u64);`, `Simd2<Bar>` has the same
+representation as `Simd2<u64>`.
+
+The `repr(simd)` may not enforce that any trait bounds exists/does the
+right thing at the type checking level for generic `repr(simd)`
+types. As such, it will be possible to get the code-generator to error
+out (ala the old `transmute` size errors), however, this shouldn't
+cause problems in practice: libraries wrapping this functionality
+would layer type-safety on top (i.e. generic `repr(simd)` types would
+use some `unsafe` trait as a bound that is designed to only be
+implemented by types that will work).
+
+It is illegal to take an internal reference to the fields of a
+`repr(simd)` type, because the representation of booleans may require
+modification, so that booleans are bit-packed. The official external
+library providing SIMD support will have private fields so this will
+not be generally observable.
+
+Adding `repr(simd)` to a type may increase its minimum/preferred
+alignment, based on platform behaviour. (E.g. x86 wants its 128-bit
+SSE vectors to be 128-bit aligned.)
+
+## Operations
+
+CPU vendors usually offer "standard" C headers for their CPU specific
+operations, such as [`arm_neon.h`][armneon] and [the `...mmintrin.h` headers for
+x86(-64)][x86].
+
+[armneon]: http://infocenter.arm.com/help/topic/com.arm.doc.ihi0073a/IHI0073A_arm_neon_intrinsics_ref.pdf
+[x86]: https://software.intel.com/sites/landingpage/IntrinsicsGuide
+
+All of these would be exposed as compiler intrinsics with names very
+similar to those that the vendor suggests (only difference would be
+some form of manual namespacing, e.g. prefixing with the CPU target),
+loadable via an `extern` block with an appropriate ABI. This subset of
+intrinsics would be on the path to stabilisation (that is, one can
+"import" them with `extern` in stable code), and would not be exported
+by `std`.
+
+Example:
+
+```rust
+extern "platform-intrinsic" {
+    fn x86_mm_abs_epi16(a: Simd8<i16>) -> Simd8<i16>;
+    // ...
+}
+```
+
+These all use entirely concrete types, and this is the core interface
+to these intrinsics: essentially it is just allowing code to exactly
+specify a CPU instruction to use. These intrinsics only actually work
+on a subset of the CPUs that Rust targets, and are only be available
+for `extern`ing on those targets. The signatures are typechecked, but
+in a "duck-typed" manner: it will just ensure that the types are SIMD
+vectors with the appropriate length and element type, it will not
+enforce a specific nominal type.
```

Comment:
as I noted above, this duck-typing may be useful for other intrinsics.  But maybe I am just being overly optimistic and/or aggressive. 

(I basically am just idly thinking about a [recent PR](https://github.com/rust-lang/rust/pull/27215) I wrote with a [commit](https://github.com/pnkfelix/rust/commit/75da81e78231a6fb5d90b3918bebe867249eb14c) that changed the signature of `move_val_init` from taking `&mut T` to taking `*mut T`, and wondering if some pain would have been avoided if the intrinsic system were more flexible in this regard. Its probably a bad idea.)